// nextra@3.1.0/dist/server/constants.js
import path from "path";
import {
  MARKDOWN_EXTENSION_REGEX,
  ERROR_ROUTES,
  DEFAULT_LOCALE
} from "../constants.js";
const CWD = process.cwd();
const PUBLIC_DIR = path.join(CWD, "public");
const CHUNKS_DIR = path.join(CWD, ".next", "static", "chunks");
const MARKDOWN_URL_EXTENSION_REGEX = /\.mdx?(?:(?=[#?])|$)/;
const IS_PRODUCTION = process.env.NODE_ENV === "production";
const DEFAULT_CONFIG = {
  staticImage: true,
  search: {
    codeblocks: true
  },
  codeHighlight: true,
  autoImportThemeStyle: true
};
const OFFICIAL_THEMES = ["nextra-theme-docs", "nextra-theme-blog"];
const META_REGEX = /_meta\.[jt]sx?$/;
const MARKDOWN_EXTENSIONS = ["md", "mdx"];
const EXTERNAL_URL_REGEX = /^https?:\/\//;
const DEFAULT_LOCALES = [""];
const DEFAULT_PROPERTY_PROPS = {
  type: "Property",
  kind: "init",
  method: false,
  shorthand: false,
  computed: false
};
const TOC_HEADING_REGEX = /^h[2-6]$/;
export {
  CHUNKS_DIR,
  CWD,
  DEFAULT_CONFIG,
  DEFAULT_LOCALE,
  DEFAULT_LOCALES,
  DEFAULT_PROPERTY_PROPS,
  ERROR_ROUTES,
  EXTERNAL_URL_REGEX,
  IS_PRODUCTION,
  MARKDOWN_EXTENSIONS,
  MARKDOWN_EXTENSION_REGEX,
  MARKDOWN_URL_EXTENSION_REGEX,
  META_REGEX,
  OFFICIAL_THEMES,
  PUBLIC_DIR,
  TOC_HEADING_REGEX
};
